sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java

```diff
@@ -765,13 +785,53 @@ static ParsedNodes create(final SqlNode node) throws ValidationException
       if (query.getKind() == SqlKind.INSERT) {
         insert = (SqlInsert) query;
         query = insert.getSource();
+
+        // Processing to be done when the original query has either of the PARTITION BY or CLUSTER BY clause
+        if (insert instanceof DruidSqlInsert) {
+          DruidSqlInsert druidSqlInsert = (DruidSqlInsert) insert;
+
+          ingestionGranularity = druidSqlInsert.getPartitionBy();
+
+          if (druidSqlInsert.getClusterBy() != null) {
+            // If we have a CLUSTER BY clause, extract the information in that CLUSTER BY and create a new SqlOrderBy
+            // node
+            SqlNode offset = null;
+            SqlNode fetch = null;
+            SqlNodeList orderByList = null;
+
+            if (query instanceof SqlOrderBy) {
+              SqlOrderBy sqlOrderBy = (SqlOrderBy) query;
+              // Extract the query present inside the SqlOrderBy (which is free of ORDER BY, OFFSET and FETCH clauses)
+              query = sqlOrderBy.query;
+
+              offset = sqlOrderBy.offset;
+              fetch = sqlOrderBy.fetch;
+              orderByList = sqlOrderBy.orderList;
+              // If the orderList is non-empty (i.e. there existed an ORDER BY clause in the query) and CLUSTER BY clause
+              // is also non-empty, throw an error
+              if (!(orderByList == null || orderByList.equals(SqlNodeList.EMPTY))
+                  && druidSqlInsert.getClusterBy() != null) {
+                throw new ValidationException(
+                    "Cannot have both ORDER BY and CLUSTER BY clauses in the same INSERT query");
+              }
+            }
+            // Creates a new SqlOrderBy query, which may have our CLUSTER BY overwritten
+            query = new SqlOrderBy(
```

Comment:
I am wondering if the SQL layer is the right place to do this transformation or should we just leave it to the native layer to use orderBy and clusterBy together. My assumption for this transformation is that we would like the query results to be ordered on same dimensions that we want to use to arrange data in segments. Maybe it's something that QueryMaker should decide?

Reply:
Currently, whatever you do enter in the "OrderBy" is getting converted as an ordering spec. I think the aim of the clause is to add syntactic sugar rather than add a different semantic.
(I am unfamiliar with Hive) I looked up Hive, which implements both of these clauses [stackoverflow](https://stackoverflow.com/questions/13715044/hive-cluster-by-vs-order-by-vs-sort-by), and it seems that both of them try and achieve similar functionality albeit with different results and different semantics. 
But yeah then it makes sense to implement both the CLUSTER BY and ORDER BY clauses and punt the actual logic to QueryMaker or a DruidRel. 
The trouble I see implementing it is 
a) We will need to convert it to a RelNode in SqlToRelConverter, and then pass it down the query stack. (This is because `CLUSTER BY` can be a reference to a column in the statement, and we probably need to use it). I am unsure if this is possible without making changes to the SqlToRelConverter itself. 
b) A new Query of type Query.INSERT would need to be created to accommodate "ClusteringSpec" in addition to the "OrderingSpec".

Reply:
I think it's OK for now to continue using the native ordering specs to represent cluster by, and to disallow having ORDER BY and CLUSTER BY at the same time. Also, IMO, we should _not_ allow ORDER BY on the SELECT for an INSERT. The error could be something like "Cannot have ORDER BY on an INSERT query; use CLUSTER BY instead."

That way, we can keep discussing if we should allow ORDER BY and CLUSTER BY to be used together, and if so, what should happen. Without blocking this PR.

Reply:
As a counter argument, it may be good to do as much semantic analysis as possible in the SQL layer. In the fullness of time, this allows us to leverage the SQL logical plan without the conversion to native. (There is some prototyping happening in that area.) The traditional approach is to do all semantic analysis in the SQL layer so that the logical plan is validated prior to the next step (in this case, prior to generating a native query.)

The one caveat, of course, is if the information is (not yet) available to the SQL layer.